### PR TITLE
restore ability to run against secured etcd

### LINKED ISSUE #18760: Restore ability to run against secured etcd
https://github.com/kubernetes/kubernetes/pull/18383/files#r47793705 removed the `--etcd-config` option from kube-apiserver, which breaks backwards compatibility and removes the ability to run against a secured etcd server.

Ideally, the etcd client package would provide a `ConfigFromFile` helper that would do the deserialize to support legacy config files, but even if they don't, we need to.

@timothysc @xiang90

### PR DESCRIPTION
**Pull request by adohe-zz**
Fixes #18760

### CODE REVIEW

**Comment by AdoHe** [comment]
@lavalamp First I want to say I am sorry about opening this PR without carefully test since I have some questions and just want to confirm: 1) why we keep `EtcdQuorumRead` in `ServerRunOption` while put other etcd related attributes in `APIServer` struct? 2) After restore etcd secure ability, there are totally 7 fields about etcd configuration in `APIServer`, also this make `newEtcdFunc` signature two long, how about refactoring these 7 fields to one struct, for example `EtcdConfig`(three are already one) 3) for now all of etcd storage tests are with insecure etcd server, shall we add the same test cases but with secure enable? 4) finally I just want to ask something about e2e test setting. I run below commands to do e2e tests, but just failed.
```
sudo ./hack/local-up-cluster.sh
make all
_output/local/bin/linux/amd64/e2e.test --host="127.0.0.1:8080" --provider="local" --ginkgo.v=true -ginkgo.dryRun=true --kubeconfig="/home/tony/.kubernetes_auth" --repo-root="/home/tony/Documents/go/src/k8s.io/kubernetes/"
```
Also the `.kubernetes_auth` has such content:
```
{
  "User": "root",
  "Password": ""
}
```
And the e2e result is:
```
------------------------------
SchedulerPredicates [Serial] 
  validates MaxPods limit number of pods that are allowed to run [Slow]
  /home/tony/Documents/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/e2e/scheduler_predicates.go:278
•
------------------------------
Pods 
  should get a host IP [Conformance]
  /home/tony/Documents/go/src/k8s.io/kubernetes/_output/local/go/src/k8s.io/kubernetes/test/e2e/pods.go:227
•
Ran 173 of 235 Specs in 0.016 seconds
SUCCESS! -- 0 Passed | 0 Failed | 0 Pending | 62 Skipped PASS
```
It seems nothing happen.
Just hope to get your feedback.

File: pkg/storage/etcd/etcd_helper.go [hunk 1]
```diff
@@ -56,26 +61,59 @@ func (c *EtcdConfig) GetType() string {
...
 }
 
+func (c *EtcdConfig) getEtcdClientConfig() etcd.Config {
+	return etcd.Config{
+		Endpoints: c.ServerList,
+		Transport: c.newHttpTransport(),
+	}
+}
+
+func (c *EtcdConfig) newHttpTransport() etcd.CancelableTransport {
```
Comment (timothysc) [inline pkg/storage/etcd/etcd_helper.go]: @xiang90 care to comment here.

File: cmd/kube-apiserver/app/server.go [hunk 2]
```diff
@@ -127,7 +130,7 @@ func newEtcd(etcdServerList []string, ns runtime.NegotiatedSerializer, storageGr
 }
 
 // parse the value of --etcd-servers-overrides and update given storageDestinations.
-func updateEtcdOverrides(overrides []string, storageVersions map[string]string, prefix string, quorum bool, storageDestinations *genericapiserver.StorageDestinations, newEtcdFn newEtcdFunc) {
+func updateEtcdOverrides(overrides []string, storageVersions map[string]string, prefix string, keyFile, certFile, caFile string, quorum bool, storageDestinations *genericapiserver.StorageDestinations, newEtcdFn newEtcdFunc) {
```
Comment (timothysc) [inline cmd/kube-apiserver/app/server.go]: In general, these signatures need refactoring, perhaps add a TODO above.

File: cmd/kube-apiserver/app/server.go [hunk 3]
```diff
@@ -87,9 +87,9 @@ func verifyClusterIPFlags(s *options.APIServer) {
 }
 
 // For testing.
-type newEtcdFunc func([]string, runtime.NegotiatedSerializer, string, string, string, bool) (storage.Interface, error)
+type newEtcdFunc func([]string, runtime.NegotiatedSerializer, string, string, string, string, string, string, bool) (storage.Interface, error)
 
-func newEtcd(etcdServerList []string, ns runtime.NegotiatedSerializer, storageGroupVersionString, memoryGroupVersionString, pathPrefix string, quorum bool) (etcdStorage storage.Interface, err error) {
+func newEtcd(etcdServerList []string, ns runtime.NegotiatedSerializer, storageGroupVersionString, memoryGroupVersionString, pathPrefix string, keyFile, certFile, caFile string, quorum bool) (etcdStorage storage.Interface, err error) {
```
Comment (timothysc) [inline cmd/kube-apiserver/app/server.go]: In general, these signatures need refactoring, perhaps add a TODO above.

Comment (AdoHe) [inline cmd/kube-apiserver/app/server.go]: Ok, will add a TODO.

Comment (lavalamp) [inline cmd/kube-apiserver/app/server.go]: Yeah, time to collect etcd creation options into a struct.

File: cmd/kube-apiserver/app/server_test.go [hunk 4]
```diff
@@ -100,7 +100,7 @@ func TestUpdateEtcdOverrides(t *testing.T) {
 	}
 
 	for _, test := range testCases {
-		newEtcd := func(serverList []string, _ runtime.NegotiatedSerializer, _, _, _ string, _ bool) (storage.Interface, error) {
+		newEtcd := func(serverList []string, _ runtime.NegotiatedSerializer, _, _, _ string, _, _, _ string, _ bool) (storage.Interface, error) {
```
Comment (lavalamp) [inline cmd/kube-apiserver/app/server_test.go]: Might be worth testing that the parameters are properly passed to this function.

File: pkg/storage/etcd/etcd_helper.go (hunk 1 above)
Comment (lavalamp) [inline pkg/storage/etcd/etcd_helper.go]: A big TODO for a test. I'm not sure the best way. I'm assuming you manually tried it and it works? :)

**Comment by lavalamp** [comment]
Thank you for throwing this together.

File: pkg/storage/etcd/etcd_helper.go (hunk 1 above)
Comment (AdoHe) [inline pkg/storage/etcd/etcd_helper.go]: @lavalamp of course, I am setting up a real test cluster to test this.

File: cmd/kube-apiserver/app/server_test.go (hunk 4 above)
Comment (AdoHe) [inline cmd/kube-apiserver/app/server_test.go]: @lavalamp OK, will do parameters check. Besides would you mind telling me what's wrong with my e2e test setting? I just can't run the e2e test.

**Comment by lavalamp** [comment]
Sorry, totally didn't see your comment with the question. I'm not sure what's wrong, but you can try adding the -ginkgo.focus=<regexp> parameter to get it to run a specific test.

File: pkg/storage/etcd/etcd_helper.go [hunk 5]
```diff
@@ -93,10 +137,10 @@ func NewEtcdStorage(client etcd.Client, codec runtime.Codec, prefix string, quor
 
 // etcdHelper is the reference implementation of storage.Interface.
 type etcdHelper struct {
-	etcdclient etcd.Client
-	client     etcd.KeysAPI
-	codec      runtime.Codec
-	copier     runtime.ObjectCopier
+	mapi   etcd.MembersAPI
```
Comment (lavalamp) [inline pkg/storage/etcd/etcd_helper.go]: etcdMembers or etcdMembersClient or membersClient. Same for 'kapi'.

(IMO: short abbreviations help readability in short function local variables and hurt it in structs as member variables)

**Comment by lavalamp** [comment]
New config struct is looking good, but I guess you have a bit more to do on this?

**Comment by AdoHe** [comment]
@lavalamp of course, I am keep going.

File: cmd/kube-apiserver/app/server_test.go [hunk 6]
```diff
@@ -100,15 +101,15 @@ func TestUpdateEtcdOverrides(t *testing.T) {
...
+		newEtcd := func(_ runtime.NegotiatedSerializer, _, _ string, etcdConfig etcdstorage.EtcdConfig) (storage.Interface, error) {
+			if !reflect.DeepEqual(test.servers, etcdConfig.ServerList) {
+				t.Errorf("unexpected server list, expected: %#v, got: %#v", test.servers, etcdConfig.ServerList)
 			}
 			return nil, nil
 		}
 		storageDestinations := genericapiserver.NewStorageDestinations()
 		override := test.apigroup + "/" + test.resource + "#" + strings.Join(test.servers, ";")
-		updateEtcdOverrides([]string{override}, storageVersions, "", false, &storageDestinations, newEtcd)
+		updateEtcdOverrides([]string{override}, storageVersions, etcdstorage.EtcdConfig{}, &storageDestinations, newEtcd)
```
Comment (lavalamp) [inline cmd/kube-apiserver/app/server_test.go]: Hm, probably shouldn't pass an empty config here?

**Comment by lavalamp** [comment]
Looking pretty good. Anything else you want to add?

File: cmd/kube-apiserver/app/server_test.go (hunk 6 above)
Comment (AdoHe) [inline cmd/kube-apiserver/app/server_test.go]: @lavalamp I will update this, also I will test this in real cluster, since we don't add any tests, I guess we still need to add tests to test secured etcd.

**Comment by timothysc** [comment]
lgtm.  

I don't want to add too much refactoring if possible, as we can always separate into another PR.

**Comment by AdoHe** [comment]
@timothysc do you have any good idea about how to test this? I am just doing local test.

**Comment by timothysc** [comment]
@AdoHe you could upload a can'd set of cert+keys, and plumb through a secure startup through the test server: https://github.com/kubernetes/kubernetes/blob/master/pkg/storage/etcd/testing/utils.go#L167

**Comment by AdoHe** [comment]
@timothysc seems a better approach I will do this test.

**Comment by lavalamp** [comment]
Cool. I think if you can do what @timothysc suggests, then we can get this in.

**Comment by AdoHe** [comment]
Of course I will add test.

**Comment by AdoHe** [comment]
@timothysc do you know how to generate the CA file? I can generate the certificate and key file, but it seems I have to set CA file when I access  a secured etcd cluster.

**Comment by timothysc** [comment]
https://github.com/openshift/openshift-ansible/tree/master/roles -> 
https://github.com/openshift/openshift-ansible/blob/master/roles/etcd_ca/tasks/main.yml

looks like we use openssl.

**Comment by AdoHe** [comment]
@lavalamp @timothysc first I want to say sorry, since I pushed some code without the unit tests pass, I just want to make sure the way I add test cases is correct. For the test it is a really big TODO, so I want to make sure I am on the right direction. Hope to get your good advices. :)

File: pkg/storage/etcd/etcd_helper.go [hunk 7]
```diff
@@ -42,64 +44,107 @@ import (
...
+			if rpool, err = util.CertPoolFromFile(c.CAFile); err != nil {
+				glog.Errorf("Error reading Etcd CA File: %v", err)
+			}
+		}
+
+		if tlsCert, err := tls.LoadX509KeyPair(c.CertFile, c.KeyFile); err == nil {
+			cc = &tls.Config{
+				RootCAs:            rpool,
+				Certificates:       []tls.Certificate{tlsCert},
+				InsecureSkipVerify: true,
```
Comment (liggitt) [inline pkg/storage/etcd/etcd_helper.go]: why is this setting insecure skip verify?

Comment (liggitt) [inline pkg/storage/etcd/etcd_helper.go]: previously, if a rootCA was specified, the connection was verified:
https://github.com/coreos/go-etcd/blob/master/etcd/client.go#L261-L296

**Comment by lavalamp** [comment]
Anything I can do to help on this one?

**Comment by AdoHe** [comment]
@lavalamp I add related test cases, but I am not sure whether this is the correct way, since we have to modify too many parts. Do you have any good idea about the test?

File: pkg/storage/etcd/etcd_helper.go (hunk 7 above)
Comment (AdoHe) [inline pkg/storage/etcd/etcd_helper.go]: Thanks @liggitt, I will update this.

File: pkg/storage/etcd/testing/utils.go [hunk 8]
```diff
@@ -164,8 +311,8 @@ func (m *EtcdTestServer) Terminate(t *testing.T) {
 }
 
 // NewEtcdTestClientServer creates a new client and server for testing
-func NewEtcdTestClientServer(t *testing.T) *EtcdTestServer {
-	server := configureTestCluster(t, "foo")
+func NewEtcdTestClientServer(t *testing.T, secure bool) *EtcdTestServer {
+	server := configureTestCluster(t, "foo", secure)
```
Comment (timothysc) [inline pkg/storage/etcd/testing/utils.go]: So I debate if there is ever a case where we don't want to default to secure for *.  @lavalamp does google support non-sec deploys?  Red Hat does not.

Comment (liggitt) [inline pkg/storage/etcd/testing/utils.go]: At the very least, I would make a second method for a secured test server to avoid ripple

Comment (AdoHe) [inline pkg/storage/etcd/testing/utils.go]: @timothysc that would be great if we enable secure by default. The tests I add are not that good, I just want to find a better way to do the test. Thanks :)

Comment (AdoHe) [inline pkg/storage/etcd/testing/utils.go]: @liggitt I did consider this way, but for some reason I decided to add a flag to control this, add a second method can eliminate great impacts on other test cases.

File: cmd/kube-apiserver/app/options/options.go [hunk 9]
```diff
@@ -220,10 +221,13 @@ func (s *APIServer) AddFlags(fs *pflag.FlagSet) {
...
 	fs.StringSliceVar(&s.EtcdServersOverrides, "etcd-servers-overrides", s.EtcdServersOverrides, "Per-resource etcd servers overrides, comma separated. The individual override format: group/resource#servers, where servers are http://ip:port, semicolon separated.")
-	fs.StringVar(&s.EtcdPathPrefix, "etcd-prefix", s.EtcdPathPrefix, "The prefix for all resource paths in etcd.")
-	fs.BoolVar(&s.EtcdQuorumRead, "etcd-quorum-read", s.EtcdQuorumRead, "If true, enable quorum read")
+	fs.StringVar(&s.EtcdConfig.Prefix, "etcd-prefix", s.EtcdConfig.Prefix, "The prefix for all resource paths in etcd.")
+	fs.StringVar(&s.EtcdConfig.KeyFile, "etcd-keyfile", s.EtcdConfig.KeyFile, "SSL key file used to secure etcd communication")
+	fs.StringVar(&s.EtcdConfig.CertFile, "etcd-certfile", s.EtcdConfig.CertFile, "SSL certification file used to secure etcd communication")
+	fs.StringVar(&s.EtcdConfig.CAFile, "etcd-cafile", s.EtcdConfig.CAFile, "SSL Certificate Authority file used to secure etcd communication")
+	fs.BoolVar(&s.EtcdConfig.Quorum, "etcd-quorum-read", s.EtcdConfig.Quorum, "If true, enable quorum read")
```
Comment (liggitt) [inline cmd/kube-apiserver/app/options/options.go]: I thought we were trying to maintain compatibility with 1.1 here... shouldn't these match the 1.1 flag where a config file is specified?

File: cmd/kube-apiserver/app/options/options.go [hunk 10]
```diff
@@ -220,10 +221,13 @@ func (s *APIServer) AddFlags(fs *pflag.FlagSet) {
...
 	fs.StringVar(&s.AdmissionControl, "admission-control", s.AdmissionControl, "Ordered list of plug-ins to do admission control of resources into cluster. Comma-delimited list of: "+strings.Join(admission.GetPlugins(), ", "))
 	fs.StringVar(&s.AdmissionControlConfigFile, "admission-control-config-file", s.AdmissionControlConfigFile, "File with admission control configuration.")
-	fs.StringSliceVar(&s.EtcdServerList, "etcd-servers", s.EtcdServerList, "List of etcd servers to watch (http://ip:port), comma separated. Mutually exclusive with -etcd-config")
+	fs.StringSliceVar(&s.EtcdConfig.ServerList, "etcd-servers", s.EtcdConfig.ServerList, "List of etcd servers to watch (http://ip:port), comma separated. Mutually exclusive with -etcd-config")
 	fs.StringSliceVar(&s.EtcdServersOverrides, "etcd-servers-overrides", s.EtcdServersOverrides, "Per-resource etcd servers overrides, comma separated. The individual override format: group/resource#servers, where servers are http://ip:port, semicolon separated.")
```
Comment (liggitt) [inline cmd/kube-apiserver/app/options/options.go]: how do the overrides interact with secured etcd?

File: cmd/kube-apiserver/app/options/options.go [hunk 11]
```diff
@@ -93,20 +93,21 @@ type APIServer struct {
 // NewAPIServer creates a new APIServer object with default parameters
 func NewAPIServer() *APIServer {
 	s := APIServer{
-		ServerRunOptions:        genericapiserver.NewServerRunOptions(),
-		APIGroupPrefix:          "/apis",
-		APIPrefix:               "/api",
-		AdmissionControl:        "AlwaysAdmit",
-		AuthorizationMode:       "AlwaysAllow",
-		DeleteCollectionWorkers: 1,
```
Comment (liggitt) [inline cmd/kube-apiserver/app/options/options.go]: this got deleted accidentally

File: pkg/storage/etcd/etcd_helper.go [hunk 12]
```diff
@@ -42,64 +44,107 @@ import (
...
 	}
-	return NewEtcdStorage(etcdClient, c.Codec, c.Prefix, c.Quorum), nil
+	return NewEtcdStorage(etcdClient, c.Codec, c.Config.Prefix, c.Config.Quorum), nil
+}
+
+// Configuration object for constructing etcd.Client
+// and accessing etcd.
+type EtcdConfig struct {
+	ServerList []string
+	Prefix     string
```
Comment (liggitt) [inline pkg/storage/etcd/etcd_helper.go]: the line between EtcdConfig and EtcdStorageConfig is unclear... I would expect Prefix in EtcdStorageConfig

File: pkg/storage/etcd/etcd_helper.go [hunk 13]
```diff
@@ -42,64 +44,107 @@ import (
...
+	}
+}
+
+func (c *EtcdConfig) newHttpTransport() etcd.CancelableTransport {
+	var cc *tls.Config = nil
+	var err error
+
+	if c.KeyFile != "" && c.CertFile != "" {
+		var rpool *x509.CertPool = nil
+		if c.CAFile != "" {
```
Comment (liggitt) [inline pkg/storage/etcd/etcd_helper.go]: you can have a CAFile without a CertFile/KeyFile... you can verify the server's certificate without providing a client certificate of your own

File: pkg/storage/etcd/testing/utils.go [hunk 14]
```diff
@@ -56,8 +66,133 @@ func newLocalListener(t *testing.T) net.Listener {
...
+		MinVersion:   tls.VersionTLS10,
+	}
+	l, err := tls.Listen("tcp", "127.0.0.1:0", &config)
+	if err != nil {
+		t.Fatal(err)
+	}
+	return l
+}
+
+func generateTestCertificates(certFile, keyFile string) error {
```
Comment (liggitt) [inline pkg/storage/etcd/testing/utils.go]: remove if unused

**Comment by AdoHe** [comment]
Thanks @liggitt I will keep working on this.

**Comment by AdoHe** [comment]
Three problems left:
1. how to keep backward compatibility with 1.1
2. how do the overrides interact with secured etcd?
3. e2e tests update to test secured etcd

File: cmd/kube-apiserver/app/options/options.go (hunk 10 above)
Comment (lavalamp) [inline cmd/kube-apiserver/app/options/options.go]: You'd have to set up all your etcd clusters to accept the same client cert for the overrides to work.

**Comment by lavalamp** [comment]
I'm OK with making people change their config if necessary, although it would be handy to do as @liggitt says and go back to a config file.

I answered the question about overrides.

I don't know if we really need an e2e test. Can we do this with an integration test? Or a unit test would be even better, @timothysc might know how to set that up?

**Comment by AdoHe** [comment]
I have already add unit test cases, @timothysc ptal. I think integration tests are necessary, will add asap. Also I will update overrides support. Old config support will be left the last thing.

File: cmd/kube-apiserver/app/options/options.go [hunk 15]
```diff
@@ -93,20 +93,21 @@ type APIServer struct {
...
-		MasterServiceNamespace:  api.NamespaceDefault,
-		RuntimeConfig:           make(util.ConfigurationMap),
-		StorageVersions:         registered.AllPreferredGroupVersions(),
-		DefaultStorageVersions:  registered.AllPreferredGroupVersions(),
+		ServerRunOptions:  genericapiserver.NewServerRunOptions(),
+		APIGroupPrefix:    "/apis",
+		APIPrefix:         "/api",
+		AdmissionControl:  "AlwaysAdmit",
+		AuthorizationMode: "AlwaysAllow",
+		EnableLogsSupport: true,
```
Comment (timothysc) [inline cmd/kube-apiserver/app/options/options.go]: You need to re-add the ```DeleteCollectionWorkers: 1,```

File: docs/admin/kube-apiserver.md [hunk 16]
```diff
@@ -66,7 +66,10 @@ kube-apiserver
       --cloud-config="": The path to the cloud provider configuration file.  Empty string for no configuration file.
       --cloud-provider="": The provider for cloud services.  Empty string for no provider.
       --cors-allowed-origins=[]: List of allowed origins for CORS, comma separated.  An allowed origin can be a regular expression to support subdomain matching.  If this list is empty CORS will not be enabled.
-      --delete-collection-workers=1: Number of workers spawned for DeleteCollection call. These are used to speed up namespace cleanup.
+      --delete-collection-workers=0: Number of workers spawned for DeleteCollection call. These are used to speed up namespace cleanup.
```
Comment (timothysc) [inline docs/admin/kube-apiserver.md]: please revert this change should still be 1

File: pkg/storage/etcd/etcd_helper_test.go [hunk 17]
```diff
@@ -352,17 +353,6 @@ func TestSetWithVersion(t *testing.T) {
 	}
 }
 
-func TestSetNilOutParam(t *testing.T) {
```
Comment (timothysc) [inline pkg/storage/etcd/etcd_helper_test.go]: why was this removed?

File: pkg/storage/etcd/etcd_watcher_secure_test.go [hunk 18]
```diff
@@ -0,0 +1,260 @@
...
+	"k8s.io/kubernetes/pkg/api/testapi"
+	"k8s.io/kubernetes/pkg/storage"
+	"k8s.io/kubernetes/pkg/storage/etcd/etcdtest"
+	etcdtesting "k8s.io/kubernetes/pkg/storage/etcd/testing"
+	"k8s.io/kubernetes/pkg/watch"
+
+	"golang.org/x/net/context"
+)
+
+func TestSecuredWatch(t *testing.T) {
```
Comment (timothysc) [inline pkg/storage/etcd/etcd_watcher_secure_test.go]: My point earlier was why not just default ```NewEtcdTestClientServer``` to always be secure, then you could eliminate the duplicate set of tests.  I can't think of a good reason not to default all tests to secure, and you would get far more coverage, to the point where I would not be concerned about integration.  In general we need to tackle integration at a later date to use the same infra.

File: cmd/kube-apiserver/app/options/options.go (hunk 10 above)
Comment (AdoHe) [inline cmd/kube-apiserver/app/options/options.go]: Now `etcd-servers-overrides` has such format: `extensions/deployments#http://ip1:port1;http://ip2:port2`, if we want to support certificates override, this format can get quite complicated, not that user friendly, I agree with @lavalamp you would have to set up all your etcd clusters to accept the same client cert.

**Comment by AdoHe** [comment]
@timothysc @lavalamp ptal.

**Comment by AdoHe** [comment]
@lavalamp @timothysc kindly ping.

**Comment by timothysc** [comment]
lgtm

File: pkg/storage/etcd/etcd_certificates/ca.pem [hunk 19]
```diff
@@ -0,0 +1,19 @@
+-----BEGIN CERTIFICATE-----
```
Comment (lavalamp) [inline pkg/storage/etcd/etcd_certificates/ca.pem]: These test files need to be in a /test/ directory. e.g. test_etcd_certificates/

File: pkg/storage/etcd/testing/utils.go [hunk 20]
```diff
@@ -68,9 +107,14 @@ func configureTestCluster(t *testing.T, name string) *EtcdTestServer {
 		t.Fatal(err)
 	}
 
-	cln := newLocalListener(t)
+	// This is a hack to get the location of certificate files.
+	_, filename, _, _ := runtime.Caller(1)
+	m.CertFile = path.Join(path.Dir(filename), "../etcd_certificates/etcdcert.pem")
```
Comment (lavalamp) [inline pkg/storage/etcd/testing/utils.go]: It might be easier to just embed the certs in go source files, instead of loading from disk.

Comment (liggitt) [inline pkg/storage/etcd/testing/utils.go]: +1  for embedding, writing to temp files at the beginning of the test, and deferring a cleanup

File: pkg/storage/etcd/etcd_certificates/ca.pem (hunk 19 above)
Comment (timothysc) [inline pkg/storage/etcd/etcd_certificates/ca.pem]: +1

**Comment by AdoHe** [comment]
just update the test according to @liggitt suggestion, thanks very much. @lavalamp ptal.

**Comment by bgrant0607** [comment]
If this is going to be in 1.2.0-beta.1, we need it wrapped up ASAP. @liggitt PTAL

File: cmd/kube-apiserver/app/server.go [hunk 21]
```diff
@@ -157,11 +155,12 @@ func updateEtcdOverrides(overrides []string, storageVersions map[string]string,
 		}
 
 		servers := strings.Split(tokens[1], ";")
+		etcdConfig.ServerList = servers
```
Comment (liggitt) [inline cmd/kube-apiserver/app/server.go]: if we're reusing the struct inside the overrides loop, shouldn't we work with a copy inside each loop, rather than stomping the server list each time through?

File: pkg/storage/etcd/testing/certificates.go [hunk 22]
```diff
@@ -0,0 +1,67 @@
...
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package testing
+
+const CAFileContent = `
```
Comment (liggitt) [inline pkg/storage/etcd/testing/certificates.go]: follow up adding comments to this with the commands used to generate the certs is helpful so we can regen when they expire

File: pkg/storage/etcd/etcd_helper.go [hunk 23]
```diff
@@ -37,69 +37,110 @@ import (
...
+	return cli, nil
+}
+
+func (c *EtcdConfig) newHttpTransport() (*http.Transport, error) {
+	info := transport.TLSInfo{
+		CertFile: c.CertFile,
+		KeyFile:  c.KeyFile,
+		CAFile:   c.CAFile,
+	}
+	cfg, err := info.ClientConfig()
```
Comment (liggitt) [inline pkg/storage/etcd/etcd_helper.go]: `info.ClientConfig()` gives no way to set `InsecureSkipVerify` to true.

In 1.1, if no CA was specified in the etcd client config, `InsecureSkipVerify` defaulted to true (https://github.com/coreos/go-etcd/blob/master/etcd/client.go#L216). I don't really think that's a good default, but there needs to be some way to accomplish that, which may mean a new field and flag (with validation that doesn't let you set both the CA file and etcd-insecure-skip-verify)

**Comment by liggitt** [comment]
just the comments on overwriting serverlist and the insecure-skip-tls-verify question.

**Comment by AdoHe** [comment]
I will fix it right now.

**Comment by liggitt** [comment]
though, if the TLSInfo used by etcd doesn't provide for insecure skip tls verify, should we? @timothysc @lavalamp @xiang90 opinions on that?

**Comment by liggitt** [comment]
The serverlist override is the only bit I would block on. InsecureSkipTLSVerify is questionable in a secured environment anyway, and could be a follow up, imo.

**Comment by AdoHe** [comment]
@liggitt I just update this, ptal.

**Comment by liggitt** [comment]
change LGTM. can you open a follow-up issue to decide on insecureSkipVerifyTLS? I guess if we ship 1.2 without etcd config file support, there's little point to re-adding it later (especially if the format isn't identical to 1.1)

**Comment by liggitt** [comment]
@bgrant0607 do you know which files the old-docs label is referring to, and if either of the doc files need reverting in this PR?

**Comment by AdoHe** [comment]
@liggitt thanks for the `release-note-breaking-change` label, we need to add some release notes, many users want to enable secured etcd connection.

**Comment by bgrant0607** [comment]
@liggitt It's docs/admin/kube-apiserver.md. @eparis's munger snags generated docs, also. However, it appears to be smart enough not to apply do-not-merge, so this is fine.

**Comment by bgrant0607** [comment]
Is this ready to LGTM @liggitt ? I see a comment above that suggests that it is.

**Comment by liggitt** [comment]
Yes, LGTM

**Comment by AdoHe** [comment]
seems the rkt unit test failed. Besides I opened a follow-up issue to decide on InsecureSkipVerifyTLS here #22858

**Comment by eparis** [comment]
@bgrant0607 is right.  kind/old-docs is pretty meaningless at this point. I might get rid of it.

**Comment by bgrant0607** [comment]
I'm fine with that @eparis

**Comment by bgrant0607** [comment]
@k8s-bot unit test this issue: #22856

**Comment by AdoHe** [comment]
seems ready to merge. @bgrant0607

**Comment by bgrant0607** [comment]
e2e flaked again. Manually merging @k8s-oncall

**Comment by eparis** [comment]
This PR is included in #22874 which is slated to be included in the release-1.2 branch.
Please verify that the cherry-pick in that PR looks correct.
